src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingContext.cs

```diff
@@ -121,62 +306,76 @@ private string CreateVaryByCacheKey(CachedVaryBy varyBy)
 
             if (cacheEntry is CachedResponse)
             {
-                // TODO: Compare cached request headers
-
-                // TODO: Content negotiation if there are multiple cached response formats?
-
-                // TODO: Verify content freshness, or else re-validate the data?
-
                 var cachedResponse = cacheEntry as CachedResponse;
+                var cachedResponseHeaders = new ResponseHeaders(cachedResponse.Headers);
 
-                var response = HttpContext.Response;
-                // Copy the cached status code and response headers
-                response.StatusCode = cachedResponse.StatusCode;
-                foreach (var pair in cachedResponse.Headers)
+                _responseTime = DateTimeOffset.UtcNow;
+                var age = _responseTime - cachedResponse.Created;
+                age = age > TimeSpan.Zero ? age : TimeSpan.Zero;
+
+                if (EntryIsFresh(cachedResponseHeaders, age, verifyAgainstRequest: true))
                 {
-                    response.Headers[pair.Key] = pair.Value;
+                    if (ValidateEncoding(cachedResponseHeaders))
+                    {
+                        var response = HttpContext.Response;
+                        // Copy the cached status code and response headers
+                        response.StatusCode = cachedResponse.StatusCode;
+                        foreach (var header in cachedResponse.Headers)
+                        {
+                            response.Headers.Add(header);
+                        }
+
+                        response.Headers[HeaderNames.Age] = ((int)Math.Min(age.TotalSeconds, int.MaxValue)).ToString(CultureInfo.InvariantCulture);
+
+                        if (_responseType == ResponseType.HeadersOnly)
+                        {
+                            response.ContentLength = 0;
```

Comment:
Don't set Content-Length = 0 for HEAD requests, they should get the real content-length.